sdk/aqueduct/dag.py

```diff
@@ -210,10 +217,20 @@ def add_artifacts(self, artifacts: List[Artifact]) -> None:
         for artifact in artifacts:
             self.artifacts[str(artifact.id)] = artifact
 
-    def update_operator(self, op: Operator) -> None:
-        """Blind replace of an operator in the dag."""
-        self.operators[str(op.id)] = op
-        self.operator_by_name[op.name] = op
+    def update_operator_spec(self, name: str, spec: OperatorSpec) -> None:
+        """Replaces an operator's spec in the dag.
+
+        The assumption validated within the method is that the caller has already validated
+        both that the operator exists, and that the spec type will be unchanged.
+        """
+        assert name in self.operator_by_name, "Operator %s does not exist." % name
+        op = self.operator_by_name[name]
+        assert get_operator_type(op) == get_operator_type_from_spec(
+            spec
+        ), "New spec has a different type."
+
+        self.operators[str(op.id)].spec = spec
+        self.operator_by_name[op.name].spec = spec
```

Comment:
This seemed safer than the blind replace of operators we currently have. Plus, we can add additional validation on the operation.

Reply:
What's the main downside of direct replacement?

Reply:
It's just bit uglier from the callers side, since editing an operator always requirements a fetch and replace. With this we can just construct the new spec from scratch. Replacing an operator fully might be better in the long term though, not sure.